standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java

```diff
@@ -1127,6 +1127,8 @@ public enum ConfVars {
       "command should handle partition retention. If enabled, and if a specific partition's age exceeded\n" +
       "retention period the partition will be dropped along with data"),
 
+    MATERIALIZED_VIEW_ENABLED("metastore.materialized.view.enabled", "metastore.materialized.view.enabled", true,
```

Comment:
If someone reaches directly reaches to HMS, invoking the HMS api's our purpose will get defeated and he would be able to create a MV.
But if we do the checks at HMS, there won't be no hacks to create a MV,

Reply:
To be honest I don't see the reason why we need this config. Don't create MV if you do not need one. 
Hive has plenty of config params already, it's really hard to navigate, and introducing a new feature flag just makes the situation worse.

Reply:
That's definitely not an HMS's responsibility to limit the engine functionality.

Reply:
@deniskuzZ @kasakrisz There are some big hive users around the globe who are using hive, but they have strict restrictions that is audited by several 3rd party providers. They always need to touch the file system when they have a query to make a log entry who touched what and when. They want to avoid any possibility to not log something. To them it would be great if they can turn it off the MV by a config. We can say do not use, but if there is an option to force to not use it is a safest for those type of hive users. 

Ps: I'm not pushing this, and I do not want to change the upstream community decision. I just wanted to give context about the use case.

-Attila

Reply:
> That's definitely not an HMS's responsibility to limit the engine functionality. What if we have Impala and Hive and just 1 engine doesn't support MV, would we disable it globally?

It has a good point. For this use case the HMS is the tight place, because they want to disable it globally. But it has sense

Reply:
> They always need to touch the file system when they have a query to make a log entry who touched what and when. 

MV creation is technically a table creation with some extra metadata. The same API is called and each time a table/mv is created a new entry is inserted into the `NOTIFICATION_LOG` table:

In this example `emp` is a table `mv1` is a materialized view.

```
NL_ID|EVENT_ID|EVENT_TIME|EVENT_TYPE                |CAT_NAME|DB_NAME           |TBL_NAME                
  304|     303|1671096468|ALLOC_WRITE_ID_EVENT      |        |default           |emp                      
  305|     304|1671096468|ALTER_TABLE               |hive    |default           |emp                      
  306|     308|1671096472|CREATE_TABLE              |hive    |default           |mv1                      
```

Could this be used for audit?

Reply:
Notification LOG only logs metadata 'write' operations, it can't be use for Read auditing. 
Second, the use case revolve around auditing the actual 'data' at the 'file level' with the user details and lot more present as part of HDFS(as for now) & some bits around Ranger Audit logs as well, MV will not bother the actual table data in the storage layer AFAIK and hence get skipped from being audited.

On a lighter note, the defaults are in a way that existing functionality for anyone won't get impacted, but if someone has usecase he can use it :-)

Reply:
if there is a lack of auditing for MV, use regular tables. Note, permissions are checked in HS2 at compile time just like for regular tables. Maybe it would be better to restrict MV creation via the Ranger rules?

Reply:
Also at HMS level we have various [listeners](https://github.com/apache/hive/blob/2cf058bbf3613a6522c674d12cd415aeb6ce8fe7/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java#L317) that are configurable. Can't we enforce checks or take additional actions there?

Reply:
@deniskuzZ - There is actually no way to differentiate in ranger if the current object is a materialized view or not. Currently, we are sending Table_or_View as a type for authorization.

Reply:
Agree that disabling the MV creation globally is somewhat drastic.  MVs (and indexes in RDBMS) are entities which privileged users (e.g DBAs) should always be allowed to create.  As previously mentioned, Create MV is almost the same as doing a CTAS which also makes a secondary copy of the data but we don't prevent people from doing CTAS. 

If we want to prevent read operations of MVs, one option is to set `hive.materializedview.rewriting` and `hive.materializedview.rewriting.sql` to false which will prevent the MVs to be used for query rewrites.  
The other suggestion about using Ranger to impose restrictions on who can do MV creation also seems reasonable and cleaner solution.  @saihemanth-cloudera  while currently there may not be a way to distinguish in Ranger, if one were to do this approach it would mean adding a new resource type for MVs in Ranger.